metaflow_custom/plugins/install_step_decorator.py

```diff
@@ -0,0 +1,57 @@
+"""Implements a step decorator to install a project package if not running locally.
+
+Implementation notes:
+- StepDecorator.task_pre_step runs just before a step begins executing in the
+  tasktime environment (not the runtime environment) - e.g. the function will
+  run on batch, not on the machine orchestrating the batch run.
+- Flow path is used as a starting point to search for setup.py file. It is
+  inferred by inspecting the arguments to the current process, the first
+  argument being the flow path.
+- `fail_loudly` defaults to false because ProjectEnvironment adds it to
+  every step, and as ProjectEnvironment is intended to replace the
+  default environment, we may not always be in a project context.
+"""
+import sys
+from pathlib import Path
+
+from metaflow.decorators import StepDecorator
+
+from metaflow_custom.utils import install_setup_py, running_in_local_env, up_to_setup_py
+
+
+class InstallProjectStepDecorator(StepDecorator):
```

Comment:
It will also be used in conjunction with `@conda` (and potentially any other future decorators that are released, e.g. K8S).

I might be misunderstanding what you mean by boilerplate and absent-minded errors but the beauty is that there isn't any boilerplate...

* `metaflow_custom.config.metaflow_config.DEFAULT_ENVIRONMENT` uses `ProjectEnvironment` by default.
* `ProjectEnvironment.decospecs` automatically adds `@install_project` to every step in the flow.

Reply:
So with `metaflow_custom` installed then this decorator gets added to every step by default.

(It can be over-ridden either by setting `DEFAULT_ENVIRONMENT` env var or passing `--environment local`)

Reply:
> So with metaflow_custom installed then this decorator gets added to every step by default.

ah gotcha!